sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/executeImmediate.scala

```diff
@@ -17,190 +17,137 @@
 
 package org.apache.spark.sql.catalyst.analysis
 
-import scala.util.{Either, Left, Right}
-
-import org.apache.spark.sql.catalyst.expressions.{Alias, Expression, VariableReference}
+import org.apache.spark.sql.catalyst.expressions.{Alias, Attribute, Exists, Expression, InSubquery, ListQuery, ScalarSubquery, VariableReference}
 import org.apache.spark.sql.catalyst.parser.ParseException
-import org.apache.spark.sql.catalyst.plans.logical.{CompoundBody, LogicalPlan, SetVariable}
+import org.apache.spark.sql.catalyst.plans.logical.{ExecutableDuringAnalysis, LocalRelation, LogicalPlan, SetVariable, UnaryNode}
 import org.apache.spark.sql.catalyst.rules.Rule
 import org.apache.spark.sql.catalyst.trees.TreePattern.{EXECUTE_IMMEDIATE, TreePattern}
 import org.apache.spark.sql.connector.catalog.CatalogManager
 import org.apache.spark.sql.errors.QueryCompilationErrors
-import org.apache.spark.sql.types.StringType
 
 /**
  * Logical plan representing execute immediate query.
  *
- * @param args parameters of query
- * @param query query string or variable
+ * @param queryParam the query expression (first child)
+ * @param args parameters from USING clause (subsequent children)
  * @param targetVariables variables to store the result of the query
  */
 case class ExecuteImmediateQuery(
+    queryParam: Expression,
```

Comment:
This is neat, but we need to check if it's string type somewhere.

Reply:
Yes, we do. We also need to check (for now) that it is a constant folding.
Really we could actually evaluate the query  (and the using clauses) using nested sql() calls. And then pass the output on.
But with constant folding we're already ahead of most DBMS I know.